### PR TITLE
[FLINK-31797][runtime] Moves LeaderElection creation into HighAvailabilityServices

### PR DESCRIPTION
**Pull request by XComp**
PR order:
* https://github.com/apache/flink/pull/21742
* https://github.com/apache/flink/pull/22379
* https://github.com/apache/flink/pull/22422
* https://github.com/apache/flink/pull/22380
* https://github.com/apache/flink/pull/22623
* https://github.com/apache/flink/pull/22384
* https://github.com/apache/flink/pull/22390
* === this PR === FLINK-31797
* https://github.com/apache/flink/pull/22601
* https://github.com/apache/flink/pull/22642
* https://github.com/apache/flink/pull/22640
* https://github.com/apache/flink/pull/22656
* https://github.com/apache/flink/pull/22661

## What is the purpose of the change

Purpose of this change is to move the `LeaderElection` creation out of the `LeaderContender` into the `HighAvailabilityServices`.

## Brief change log

* Introduced `StandaloneLeaderElection`
* Added `TestingLeaderElection` to use for tests
* Moved `LeaderElectionService::createLeaderElection()` into `HighAvailabilityServices`

## Verifying this change

* All tests should still pass

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: yes
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Comment by XComp** [comment]
@flinkbot run azure

File: flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/AbstractHaServices.java [hunk 1]
```diff
@@ -240,13 +241,13 @@ public CompletableFuture<Void> globalCleanupAsync(JobID jobID, Executor executor
                 executor);
     }
 
-    private LeaderElectionService createLeaderElectionService(String leaderName) throws Exception {
+    private LeaderElection createLeaderElectionService(String leaderName) throws Exception {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/AbstractHaServices.java:244]: rename to `createLeaderElection`? Although it does a bit more...

**Review by zentol — COMMENTED** [review]
Didn't spot anything odd! Would take another final look after things were rebased.

**Comment by XComp** [comment]
I rebased the PR onto FLINK-31785 (PR #22390 ) and renamed the method in accordance to the review comment

**Comment by XComp** [comment]
Rebased PR to master after FLINK-31785 was merged.

File: flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherCleanupITCase.java [hunk 2]
```diff
@@ -157,10 +157,9 @@ public void testCleanupThroughRetries() throws Exception {
         jobGraphStore.start(NoOpJobGraphListener.INSTANCE);
         haServices.setJobGraphStore(jobGraphStore);
 
-        // Construct leader election service.
-        final TestingLeaderElectionService leaderElectionService =
-                new TestingLeaderElectionService();
-        haServices.setJobMasterLeaderElectionService(jobId, leaderElectionService);
+        // Construct leader election.
+        final TestingLeaderElection leaderElection = new TestingLeaderElection();
+        haServices.setJobMasterLeaderElectionService(jobId, leaderElection);
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherCleanupITCase.java:162]: should the method name be updated?

File: flink-runtime/src/test/java/org/apache/flink/runtime/highavailability/TestingHighAvailabilityServices.java [hunk 3]
```diff
@@ -49,20 +49,20 @@ public class TestingHighAvailabilityServices implements HighAvailabilityServices
...
 
-    private volatile Function<JobID, LeaderElectionService> jobMasterLeaderElectionServiceFunction =
+    private volatile Function<JobID, LeaderElection> jobMasterLeaderElectionServiceFunction =
             ignored -> null;
 
     private ConcurrentHashMap<JobID, LeaderRetrievalService> jobMasterLeaderRetrievers =
             new ConcurrentHashMap<>();
 
-    private ConcurrentHashMap<JobID, LeaderElectionService> jobManagerLeaderElectionServices =
+    private ConcurrentHashMap<JobID, LeaderElection> jobManagerLeaderElectionServices =
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/highavailability/TestingHighAvailabilityServices.java:58]: rename field and setter

**Review by zentol — APPROVED** [review]

**Comment by XComp** [comment]
Force pushed formatting fix

**Comment by XComp** [comment]
rebase

**Pull request merged**